### PR TITLE
helm values template

### PR DESCRIPTION
**Pull request by EItanya**

### CODE REVIEW

File: install/helm/gloo/templates/10-ingress-deployment.yaml [hunk 1]
```diff
@@ -17,8 +17,8 @@ spec:
         gloo: ingress
     spec:
       containers:
-      - image: {{ .Values.deployment.ingress.image }}
-        imagePullPolicy: {{ .Values.deployment.imagePullPolicy }}
+      - image: "{{ .Values.ingress.deployment.image.repository }}:{{ .Values.ingress.deployment.image.tag }}"
+        imagePullPolicy: {{ .Values.ingress.deployment.image.pull_policy }}
```
Comment (ilackarms) [inline install/helm/gloo/templates/10-ingress-deployment.yaml]: nit: `pullPolicy`
seems to be the conventional way of casing in kube

**Comment by ilackarms** [comment]
any reason for the change from camel to snake case?

**Comment by rickducott** [comment]
Continuing work on this PR, just to be clear the end state we should get to: 
* The helm directory contains template files for `Chart.yaml` and `values.yaml`, since they require version tags that are only known at build time. This means `Chart.yaml` is deleted and replaced with `Chart-template.yaml`, `values.yaml` is replaced with `values-template.yaml`
* `Chart.yaml` and `values.yaml` (and any other generated file) is removed and .gitignored from repo
* The build script is updated to first generate `Chart.yaml` and `values.yaml`, then build a helm chart.
* Any file that is generated from the helm chart (i.e. the kube install manifest `install.yaml`) is deleted from this repo and .gitignored. After generating the helm chart, the build produces those manifests via a helm command. 
* The release is updated to publish install manifests as part of the release artifacts. 
* The generated yaml go bytes wrappers are deleted, and the CLI is updated to by default install the published manifests. (for instance, if you download glooctl v0.x.y, then run `glooctl install kube`, it will apply the install.yaml manifest that is published during the v0.x.y release build)

The makefile should support workflows for release, where the CI build provides a tagged version, and for local development, where the make target is supplied with an explicit version.

File: Makefile [hunk 2]
```diff
@@ -279,7 +279,10 @@ RELEASE_BINARIES :=
 ifeq ($(RELEASE),"true")
 	RELEASE_BINARIES := \
```
Comment (ilackarms) [inline Makefile]: nit: rename RELEASE_BINARIES to something covering manifests, e.g. RELEASE_ARTIFACTS

File: install/helm/gloo/generate/values.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+type Config struct {
+	Namespace    Namespace `json:"namespace"`
+	Rbac         Rbac `json:"rbac"`
+	Settings     Settings `json:"settings"`
+	Gloo         Gloo `json:"gloo"`
+	Discovery    Discovery `json:"discovery"`
+	Gateway      Gateway `json:"gateway"`
+	GatewayProxy GatewayProxy `json:"gatewayProxy"`
+	Ingress      Ingress `json:"ingress"`
+	IngressProxy IngressProxy `json:"ingressProxy"`
```
Comment (ilackarms) [inline install/helm/gloo/generate/values.go]: suggestion to replace `GatewayProxy`, `IngressProxy`, and `KnativeProxy` structs with `Proxy` struct as they should be almost identical (except for their name & role in [their respective configmaps](https://github.com/solo-io/gloo/blob/master/install/helm/gloo/templates/9-gateway-proxy-configmap.yaml#L10)

File: install/helm/gloo/templates/settings.yaml [hunk 4]
```diff
@@ -12,11 +12,8 @@ spec:
   kubernetesConfigSource: {}
   kubernetesSecretSource: {}
   refreshRate: 60s
-{{- if gt (len .Values.settings.watchNamespaces) 0 }}
+{{- with .Values.settings.watchNamespaces }}
   watchNamespaces:
-{{- range .Values.settings.watchNamespaces }}
```
Comment (ilackarms) [inline install/helm/gloo/templates/settings.yaml]: i believe this becomes `{{- range . }}`

File: projects/gloo/cli/pkg/cmd/install/knative.go [hunk 5]
```diff
@@ -57,6 +64,16 @@ func KnativeCmd(opts *options.Options) *cobra.Command {
 	return cmd
 }
 
+func readKnativeManifest(version string) ([]byte, error) {
+	urlTemplate := "https://github.com/solo-io/gloo/releases/download/v%s/knative-no-istio-0.3.0.yaml"
+	return ReadManifest(version, urlTemplate)
+}
+
+func readGlooKnativeManifest(version string) ([]byte, error) {
+	urlTemplate := "https://github.com/solo-io/gloo/releases/download/v%s/gloo-knative.yaml"
```
Comment (ilackarms) [inline projects/gloo/cli/pkg/cmd/install/knative.go]: extract to consts

File: projects/gloo/cli/pkg/cmd/install/util.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Write the body to file
+	return ioutil.ReadAll(resp.Body)
+}
+
+func ReadManifest(version, urlTemplate string) ([]byte, error) {
```
Comment (ilackarms) [inline projects/gloo/cli/pkg/cmd/install/util.go]: does this function need to be public?

Comment (rickducott) [reply projects/gloo/cli/pkg/cmd/install/util.go]: Nope, made private

File: Makefile (hunk 2 above)
Comment (rickducott) [reply Makefile]: I'm going to probably move them to a separate RELEASE_YAMLS list, I don't think we need SHAs for each of the yamls (it's very busy / hard to read in the github UI)

**Comment by rickducott** [comment]
Tested make target for yamls by commenting out the binaries and running `make release TAGGED_VERSION=v0.6.6`. yamls with correct versions got uploaded to v0.6.6 release. 

Ran `make glooctl TAGGED_VERSION=v0.6.6` to produce CLI locally with proper version. Then ran `_output/glooctl install kube`, and `output/glooctl install knative` to test release path. 

Ran `make glooctl` to produce CLI locally with dev version, then ran `_output/glooctl install kube` to get error and provided `-f install/gloo.yaml` to test dev install workflow. 

Ran `make glooctl` to produce CLI locally with dev version, then ran `_output/glooctl install knative` to get error and provided `-f install/gloo-knative.yaml -knative-manifest install/integrations/knative-no-istio-0.3.0.yaml` to test dev install workflow.

**Comment by EItanya** [comment]
Approved LGTM

**Comment by rickducott** [comment]
@EItanya reviewed and approved in person, since this is his PR he can't approved so I've added approval.

**Comment by rickducott** [comment]
Latest build flaked #264

**Pull request merged**